Transform gizmo

almarklein opened the pull request:
Closes #186 

* [x] Translation per dimension.
* [x] Translation in planes.
* [x] Translate in view plane.
* [x] Scale per dimension.
* [x] Uniform scale.
* [x] Rotate per dimension.
* [x] Toggle between local and world space. *and screen space*.
* [x] Disable handles when their movement becomes ill-defined from the current view.
* [x] Make the gizmo "always on top".
* [x] Scale widget so it always has the same size in screen coords.
* [x] We update *all* the things each draw. Can we make it only update if needed? -> no need.
* [x] ~Make renderer memoize pick info.~ -> no need with new event system. Added a note.
* [x] Decide+implement a way to make gizmo and camera behave well w.r.t. events.
* [x] Turn camera back on (without having to hold shift).
* [x] Highlight  handle during interaction
* [x] Prettier colors.
* [x] Resilience for camera scaling (anisotropy and flipped axii).
* [x] Revise translate scaling.
* [x] Cleanup and self-review.
* [ ] Mouseover highlighting

![Screen Recording 2022-03-11 at 14 08 07](https://user-images.githubusercontent.com/3015475/157873077-eb6f2fcd-762c-42f5-b42b-302f123a85e4.gif)

Korijn commented on pygfx/helpers/_gizmo.py:
```diff
@@ -0,0 +1,233 @@
+
+    Aguments:
+        object_to_control (WorldObject): the object to transform with the gizmo.
+        screen_size (float): the approximate size of the widget in logical pixels.
```
is it the size of the diagonal?

almarklein replied:
No, more like the average size (in x and y).

Korijn replied:
So it's both the width and the height?

almarklein commented on pygfx/helpers/_gizmo.py:
```diff
@@ -0,0 +1,397 @@
+        pos2.multiply_scalar(2 / 100 / canvas_size[1])
+        self._screen_vecs = pos1, pos2
+
+        # Determine the scale of this object
+        avg_length = 0.5 * (
+            self._screen_vecs[0].length() + self._screen_vecs[1].length()
+        )
+        scale_scalar = self._screen_size * avg_length
+        scale = [scale_scalar, scale_scalar, scale_scalar]
```
It's this. It's a bit hard to explain :)  we transform a horizontal and vertical vector in screen space to scene space. The average of *these* vectors is used to scale the widget, so that it approximately takes up the intended screen space.

almarklein commented:
This is finally taking shape. As a sort of experiment I made it so you can change the mode by clicking on the sphere in the center. This iterates through model-space, world-space, and screen-space. I think this might work!

brisvag commented:
This looks amazing! :D @alisterburt and @kevinyamauchi,  maybe you find some inspiration for napari-threedee :)

Korijn commented:
I suggest to prioritize "Resilience for camera scaling (anisotropy and flipped axii)." from your to-do list, and wait for Berend to implement the event system until you tackle the rest of the items?

almarklein commented:
I was too curious about these new event system ideas, so I implemented a minimal ad-hoc thingy in the example, and removed the need to hold shift to use the camera. We can implement this the proper way once the event system is ready.

almarklein commented:
Movie of current state in top post. The translation now works much better that it did last time I showed it.

almarklein commented:
Interestingly, the highlighting introduced a noticeable delay. It looks like updating the uniform material buffer for all 18 elements starts to become slow. I changed the highlighting to only make the relavant line(s) a bit thicker, which also more visually appealing IMO.

Korijn commented:
> It looks like updating the uniform material buffer for all 18 elements starts to become slow.

If that's significantly slower than it is supposed to be, I suggest we open a new ticket. If you think it's only natural, then no problem.

almarklein commented:
At first I thought it might be natural, but now I don't think it is. Also, all object uniforms are updated at each draw, and there is no problem there ... #257

Korijn commented on examples/helpers_gizmo.py:
```diff
@@ -0,0 +1,47 @@
+def animate():
+    controls.update_camera(camera)
+    renderer.render(scene, camera, flush=False)
+    renderer.render(gizmo, camera, clear_color=False)
```
can you explain why it needs to be rendered separately?

almarklein replied:
Because the gizmo is rendered as an overlay. Otherwise the depth testing would hide part of it behind the object being manipulated, and other objects as well. Will add a comment.

Korijn commented on pygfx/helpers/_gizmo.py:
```diff
@@ -0,0 +1,747 @@
+            self._object_to_control = None
+        else:
+            assert isinstance(object, WorldObject)
+            self._object_to_control = object
```
how about:

```python
        if object is None or isinstance(object, WorldObject):
            self._object_to_control = object
        else:
            raise ValueError("object is not None or a WorldObject instance")
```

almarklein replied:
Good point.

Korijn commented on pygfx/helpers/_gizmo.py:
```diff
@@ -0,0 +1,747 @@
+        """
+        modes = "object", "world", "screen"
+        if not mode:
+            mode = {"object": "world", "world": "screen"}.get(self._mode, "object")
```
this is the first instance I've seen where `itertools.cycle` would actually be useful :) anyway, it's fine!

Korijn commented on pygfx/helpers/_gizmo.py:
```diff
@@ -0,0 +1,747 @@
+
+        # Create 1D translate handles
+        cone_geo = cone_geometry(0.1, 0.17)
+        cone_geo.positions.data[:] = cone_geo.positions.data[:, ::-1]
```
what does this line do?

almarklein replied:
Reverses xyz -> zyx, to put the cones in the correct orientation. Added a comment.

Korijn commented on pygfx/helpers/_gizmo.py:
```diff
@@ -0,0 +1,747 @@
+        # Reset thickness of all lines
+        for ob in self._line_children + self._arc_children:
+            if ob.material.thickness != 3:
+                ob.material.thickness = 3
```
might want to make this a constant as well?

almarklein replied:
Done.

alisterburt commented:
just chiming in to say this is beautiful @almarklein ! 😍

almarklein commented on pygfx/helpers/_gizmo.py:
```diff
@@ -0,0 +1,747 @@
+
+        # Init
+        self.set_object(object)
+        self._create_elements()
```
👍

almarklein commented on pygfx/helpers/_gizmo.py:
```diff
@@ -0,0 +1,747 @@
+        sphere_geo = sphere_geometry(0.07)
+        scale_uniform = Mesh(
+            sphere_geo,
+            MeshBasicMaterial(color="#ffffff"),
```
👍

almarklein commented on pygfx/helpers/_gizmo.py:
```diff
@@ -0,0 +1,747 @@
+    def _on_mode_switch(self):
+        """When the mode is changed, some adjustments are made."""
+
+        # Note: the visibility of certain elements is handled in _update_visibility()
```
Rephrased that comment.

berendkleinhaneveld reviewed (commented):
Looks great, also the readability!

Korijn commented:
I think we can merge this, just gotta be aware that we need to update it's event handling when we merge #258!

almarklein commented:
Yes, let's update the event handling when the new system is in place!
